habitat/plan.sh

```diff
@@ -0,0 +1,98 @@
+pkg_name=chef-client
+pkg_origin=chef
+pkg_maintainer="The Chef Maintainers <humans@chef.io>"
+pkg_description="The Chef Client"
+pkg_version=$(cat ../VERSION)
+pkg_source=nosuchfile.tar.gz
+pkg_license=('Apache-2.0')
+pkg_bin_dirs=(bin)
+pkg_build_deps=(core/make core/gcc core/coreutils core/rsync core/git)
+pkg_deps=(core/glibc core/ruby core/libxml2 core/libxslt core/libiconv core/xz core/zlib core/bundler core/openssl core/cacerts core/libffi)
+pkg_svc_user=root
+
+do_prepare() {
+  export OPENSSL_LIB_DIR=$(pkg_path_for openssl)/lib
+  export OPENSSL_INCLUDE_DIR=$(pkg_path_for openssl)/include
+  export SSL_CERT_FILE=$(pkg_path_for cacerts)/ssl/cert.pem
+
+  build_line "Setting link for /usr/bin/env to 'coreutils'"
+  [[ ! -f /usr/bin/env ]] && ln -s $(pkg_path_for coreutils)/bin/env /usr/bin/env
+
+  rsync -vaP --delete --exclude habitat --exclude results --exclude .git --exclude Gemfile.lock $PLAN_CONTEXT/../ $HAB_CACHE_SRC_PATH/$pkg_dirname
```

Comment:
In our tinkering, we were able to speed up the "get the code in place" part of the build by using `git archive` to make a tarball in the conventional location for the default `do_unpack()` to pick up to Do The Thing.

Also a benefit, no accidental packaging up of a dirty working directory.

```diff
diff --git a/habitat/plan.sh b/habitat/plan.sh
index 32c5943..7c85462 100644
--- a/habitat/plan.sh
+++ b/habitat/plan.sh
@@ -3,13 +3,47 @@ pkg_origin=chef
 pkg_maintainer="The Chef Maintainers <humans@chef.io>"
 pkg_description="The Chef Client"
 pkg_version=$(cat ../VERSION)
-pkg_source=nosuchfile.tar.gz
+pkg_source=not_downloaded
 pkg_license=('Apache-2.0')
 pkg_bin_dirs=(bin)
 pkg_build_deps=(core/make core/gcc core/coreutils core/rsync core/git)
 pkg_deps=(core/glibc core/ruby core/libxml2 core/libxslt core/libiconv core/xz core/zlib core/bundler core/openssl core/cacerts core/libffi)
 pkg_svc_user=root

+do_begin() {
+  pkg_filename=${pkg_dirname}.tar.gz
+  return 0
+}
+
+do_download() {
+  build_line "Fake download! Creating archive of latest repository commit."
+  # source is in this repo, so we're going to create an archive from the
+  # appropriate path within the repo for the rest of the plan callback chain
+  # to pick up on
+  cd $PLAN_CONTEXT/../
+  # don't include the plan and such in the source tarball produced
+  echo "habitat/ export-ignore" >> .gitattributes
+  git archive --prefix=${pkg_name}-${pkg_version}/ --output=$HAB_CACHE_SRC_PATH/${pkg_filename} HEAD
+  # set .gitattributes back to the way they were
+  git checkout .gitattributes
+}
+
+do_verify() {
+  build_line "Skipping checksum verification on the archive we just created."
+  return 0
+}
+
+do_clean() {
+  do_default_clean
+
+  # Clean up the `env` link, if we set it up.
+  if [[ `readlink /usr/bin/env` = "$(pkg_path_for coreutils)/bin/env" ]]; then
+    build_line "Removing the symlink we created for '/usr/bin/env'"
+    rm /usr/bin/env
+  fi
+  return 0
+}
+
 do_prepare() {
   export OPENSSL_LIB_DIR=$(pkg_path_for openssl)/lib
   export OPENSSL_INCLUDE_DIR=$(pkg_path_for openssl)/include
@@ -18,7 +52,6 @@ do_prepare() {
   build_line "Setting link for /usr/bin/env to 'coreutils'"
   [[ ! -f /usr/bin/env ]] && ln -s $(pkg_path_for coreutils)/bin/env /usr/bin/env

-  rsync -vaP --delete --exclude habitat --exclude results --exclude .git --exclude Gemfile.lock $PLAN_CONTEXT/../ $HAB_CACHE_SRC_PATH/$pkg_dirname
   return 0
 }

@@ -80,18 +113,12 @@ do_install() {
   fi
 }

-# Stubs
-do_verify() {
-  return 0
+do_end() {
+  do_clean
 }

-do_download() {
-  return 0
-}

-do_unpack() {
-  return 0
-}
+# Stubs

 do_strip() {
   return 0
```